test/sync/mutual_exclusion/locks/unique_lock/cons/move_ctor_upgrade_lock_for_pass.cpp

```diff
@@ -30,6 +30,7 @@ boost::upgrade_mutex m;
 
 int main()
 {
+#if defined(BOOST_THREAD_PLATFORM_PTHREAD)
```

Comment:
Now that win32 works with the pthread/shared_mutex generic implementation this should works. What was wrong?

Reply:
I'd have to go back and re-run it again to remember the exact error. It was something related to not being able to copy (or maybe move) between mutexes or locks of different types.

Reply:
Here's an example of the error that occurs in these tests with Visual Studio 2010. I'm not sure what the problem is. Do you have any ideas?

```
sync\mutual_exclusion\locks\unique_lock\cons\move_ctor_upgrade_lock_try_pass.cpp(34) : error C2664: 'boost::unique_lock<Mutex>::unique_lock<boost::try_to_lock_t>(Mutex &,const TimeDuration &)' : cannot convert parameter 1 from 'boost::upgrade_lock<Mutex>' to 'boost::upgrade_mutex &'
        with
        [
            Mutex=boost::upgrade_mutex,
            TimeDuration=boost::try_to_lock_t
        ]
        and
        [
            Mutex=boost::upgrade_mutex
        ]
```

Reply:
Oh, I see what's going on, although I'm not sure why. In `lock_types.hpp`, Visual Studio 2010 is choosing this constructor:
```
    template<typename TimeDuration>
    unique_lock(Mutex& m_,TimeDuration const& target_time);
```
instead of this constructor:
```
    unique_lock(Mutex& m_, try_to_lock_t);
```

Reply:
Oh.....in `lock_types.hpp`, this is the constructor it needs, but it's not available on Windows:

```
#if defined(BOOST_THREAD_PLATFORM_PTHREAD)
    // Conversion from upgrade locking
    unique_lock(BOOST_THREAD_RV_REF_BEG upgrade_lock<mutex_type> BOOST_THREAD_RV_REF_END ul, try_to_lock_t)
    : m(0),is_locked(false)
    {
      if (BOOST_THREAD_RV(ul).owns_lock())
      {
        if (BOOST_THREAD_RV(ul).mutex()->try_unlock_upgrade_and_lock())
        {
          m = BOOST_THREAD_RV(ul).release();
          is_locked = true;
        }
      }
      else
      {
        m = BOOST_THREAD_RV(ul).release();
      }
    }
#endif
```

Reply:
I got it...the #if defines were added by you and I just failed to remove them. That's why the tests are failing. I'll take care of it. Sorry for the spam.

Reply:
No problem. I know that this is working now in develop, so I know we just did something wrong.